README.md
# oai-ore

A Clojure library used to generate Open Archives Initiative Object Reuse and Exchange files for the CyVerse Data Commons
repository. The primary purpose of this library is to make it easier to generate OAI-ORE files from a set of URIs and
attribute metdata stored in the CyVerse Data Store.

## Usage

All examples assume that these commands have been executed in the REPL:

``` clojure
(defn- file-info-from-id [id] {:id  id :uri (str "http://foo.org/" id)})
(require '[org.cyverse.oai-ore :as ore])
(require '[clojure.data.xml :refer :all])
(def agg-uri "http://foo.org")
(def arch (file-info-from-id "bar"))
(def metadata (file-info-from-id "baz"))
(def archived-files (mapv file-info-from-id ["bar1" "bar2"]))
```

The suggested way to build an OAI-ORE file is to use the `build-ore` function. The result of this function is an
instance of `org.cyverse.oai-ore.Ore`, which can then be converted to RDF/XML and serialized. The simplest such file is
an empty archive:

``` clojure
(def empty-ore (ore/build-ore agg-uri arch []))
```

Once you have the `org.cyverse.oai-ore.Ore` instance, you can convert it to RDF/XML by calling its `to-rdf` method:

``` clojure
(def rdf (ore/to-rdf empty-ore))
```

The result of calling this method is an instance of `clojure.data.xml.Element`, which can then be serialized using any
serialization method available in `org.clojure/data.xml`. For example, you can use the following commands to emit
pretty-printed RDF/XML:

``` clojure
(print (indent-str rdf))
```

An empty OAI-ORE file is good for an example, but not very useful. It's possible to add aggregated entites by including
some file information in the third argument to `build-ore`:

``` clojure
(def populated-ore (ore/build-ore agg-uri arch archived-files))
```

This library also supports several of the metadata attributes that are used when a data set is published to the CyVerse
Data Commons repository. The following attributes are currently supported:

| CyVerse Attribute     | ORE Element           |
| --------------------- | --------------------- |
| datacite.title        | dc:title              |
| datacite.publisher    | dc:publisher          |
| datacite.creator      | dc:creator            |
| datacite.resourcetype | dc:type               |
| contributorName       | dc:contributor        |
| Subject               | dc:subject            |
| Rights                | dc:rights             |
| Description           | dc:description        |
| Identifier            | dcterms:identifier    |
| geoLocationBox        | dcterms:Box           |
| geoLocationPlace      | dcterms:Location      |
| geoLocationPoint      | dcterms:Point         |

Any attribute that is associated with the data set that is not in this list is ignored. Similarly, any attribute that is
in the list but either contains an empty value or is not associated with the data set is ignored:

``` clojure
(def attr-ore (ore/build-ore agg-uri arch archived-files [{:attr "datacite.title" :value "The Title"}
                                                          {:attr "datacite.creator" :value "The Creator"}
                                                          {:attr "ignored.attribute" :value "Who Cares?"}
                                                          {:attr "Subject" :value ""}]
                             metadata))
```

Serializing the RDF using `(print (indent-str (ore/to-rdf attr-ore)))` should produce the following output:

``` xml
<?xml version="1.0" encoding="UTF-8"?>
<rdf:RDF xmlns:dc="http://purl.org/dc/elements/1.1/"
         xmlns:owl="http://www.w3.org/2002/07/owl#"
         xmlns:foaf="http://xmlns.com/foaf/0.1/"
         xmlns:dcterms="http://purl.org/dc/terms/"
         xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
         xmlns:ore="http://www.openarchives.org/ore/terms/"
         xmlns:rdfs="http://www.w3.org/2000/01/rdf-schema#"
         xmlns:cito="http://purl.org/spar/cito/">
  <rdf:Description rdf:about="http://foo.org">
    <rdf:type rdf:resource="http://www.openarchives.org/ore/terms/Aggregation"/>
    <ore:aggregates rdf:resource="http://foo.org/baz"/>
    <ore:aggregates rdf:resource="http://foo.org/bar1"/>
    <ore:aggregates rdf:resource="http://foo.org/bar2"/>
    <dc:title>The Title</dc:title>
    <dc:creator>The Creator</dc:creator>
  </rdf:Description>
  <rdf:Description rdf:about="http://foo.org/bar">
    <dcterms:identifier>bar</dcterms:identifier>
    <rdf:type rdf:resource="http://www.openarchives.org/ore/terms/ResourceMap"/>
    <ore:describes rdf:resource="http://foo.org"/>
  </rdf:Description>
  <rdf:Description rdf:about="http://foo.org/baz">
    <dcterms:identifier>baz</dcterms:identifier>
    <cito:documents rdf:resource="http://foo.org/bar1"/>
    <cito:documents rdf:resource="http://foo.org/bar2"/>
  </rdf:Description>
  <rdf:Description rdf:about="http://foo.org/bar1">
    <dcterms:identifier>bar1</dcterms:identifier>
    <cito:isDocumentedBy rdf:resource="http://foo.org/baz"/>
  </rdf:Description>
  <rdf:Description rdf:about="http://foo.org/bar2">
    <dcterms:identifier>bar2</dcterms:identifier>
    <cito:isDocumentedBy rdf:resource="http://foo.org/baz"/>
  </rdf:Description>
</rdf:RDF>
```

Note: the namespace declarations have been reformatted for readability.

## License

http://www.cyverse.org/license


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions
of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]

## [1.0.3]
### Added
- A new arity for build-ore.
- Support for metadata files that are used to describe other files in a data set.

## [1.0.2]
### Changed
- Replaced `dc:identifier` with `dcterms:identifier` in generated files.
- All archived files, including the resource map now have a `dcterms:identifier` element.
- Metadata attributes will now be trimmed before being placed in the ORE file.

## [1.0.1]
### Added
- format-id constant.

## 1.0.0
### Added
- RdfSerializable protocol.
- Aggregation type.
- Archive type.
- ArchivedFile type.
- Ore type.
- build-ore

[Unreleased]: https://github.com/cyverse-de/oai-ore/compare/1.0.3...HEAD
[1.0.3]: https://github.com/cyverse-de/oai-ore/compare/1.0.2...1.0.3
[1.0.2]: https://github.com/cyverse-de/oai-ore/compare/1.0.1...1.0.2
[1.0.1]: https://github.com/cyverse-de/oai-ore/compare/1.0.0...1.0.1


test/org/cyverse/oai_ore_test.clj
(ns org.cyverse.oai-ore-test
  (:require [clojure.data.xml :as xml]
            [clojure.java.io :as io]
            [clojure.test :refer :all]
            [org.cyverse.oai-ore :refer :all]))

(defn- file-info-from-id [id]
  {:id  id
   :uri (str "http://foo.org/" id)})

;; URIs to use when building test OREs.
(def ^:private agg-uri "http://foo.org")
(def ^:private arch (file-info-from-id "6cf74bf2-f172-47b4-94cc-acfc82559fc6"))
(def ^:private metadata (file-info-from-id "d0b7aa9a-1c22-4a59-9b04-3f35f3a9e15d"))
(def ^:private archived-files
  (mapv file-info-from-id ["73cd1d79-28c9-4a52-9060-cf9c65905922" "6772f358-9c10-4fa0-9fec-1376816b37e4"]))

;; Comparing generated XML to parsed XML requires us to serialize and parse the generated XML.
(defn- test-ore [filename ore & [debug]]
  (when debug
    (println "Actual:" (xml/indent-str (to-rdf ore)))
    (println "Expected:" (xml/indent-str (xml/parse (io/reader (io/resource filename))))))
  (is (= (xml/parse (io/reader (io/resource filename)))
         (xml/parse-str (xml/emit-str (to-rdf ore))))))

(deftest test-empty-ore
  (testing "Empty ORE."
   (test-ore "empty-ore.rdf" (build-ore agg-uri arch []))
   (test-ore "empty-ore.rdf" (build-ore agg-uri arch [] [{:attr "datacite.title" :value ""}]))
   (test-ore "empty-ore-meta.rdf" (build-ore agg-uri arch [] [] metadata))))

(deftest test-ore-with-files
  (testing "ORE with files."
    (test-ore "ore-with-files.rdf" (build-ore agg-uri arch archived-files))
    (test-ore "ore-with-files.rdf" (build-ore agg-uri arch archived-files [{:attr "datacite.title" :value ""}]))
    (test-ore "ore-with-files-meta.rdf" (build-ore agg-uri arch archived-files [] metadata))))

(deftest test-ore-with-title
  (testing "ORE with title."
    (test-ore "ore-with-title.rdf"
              (build-ore agg-uri arch archived-files [{:attr "datacite.title" :value "The Title"}]))
    (test-ore "ore-with-title.rdf"
              (build-ore agg-uri arch archived-files [{:attr "datacite.title" :value "The Title"}
                                                      {:attr "datacite.foo" :value "Bar"}]))
    (test-ore "ore-with-title.rdf"
              (build-ore agg-uri arch archived-files [{:attr "datacite.title" :value "The Title"}
                                                      {:attr "datacite.publisher" :value ""}]))))

(deftest test-ore-with-publisher
  (testing "ORE with publisher."
    (test-ore "ore-with-publisher.rdf"
              (build-ore agg-uri arch archived-files [{:attr "datacite.publisher" :value "The Publisher"}]))
    (test-ore "ore-with-publisher.rdf"
              (build-ore agg-uri arch archived-files [{:attr "datacite.publisher" :value "The Publisher"}
                                                      {:attr "datacite.foo" :value "Bar"}]))
    (test-ore "ore-with-publisher.rdf"
              (build-ore agg-uri arch archived-files [{:attr "datacite.publisher" :value "The Publisher"}
                                                      {:attr "datacite.title" :value ""}]))))

(deftest test-ore-with-creator
  (testing "ORE with creator."
    (test-ore "ore-with-creator.rdf"
              (build-ore agg-uri arch archived-files [{:attr "datacite.creator" :value "The Creator"}]))
    (test-ore "ore-with-creator.rdf"
              (build-ore agg-uri arch archived-files [{:attr "datacite.creator" :value "The Creator"}
                                                      {:attr "datacite.foo" :value "Bar"}]))
    (test-ore "ore-with-creator.rdf"
              (build-ore agg-uri arch archived-files [{:attr "datacite.creator" :value "The Creator"}
                                                      {:attr "datacite.title" :value ""}]))))

(deftest test-ore-with-type
  (testing "ORE with type."
    (test-ore "ore-with-type.rdf"
              (build-ore agg-uri arch archived-files [{:attr "datacite.resourcetype" :value "The Type"}]))
    (test-ore "ore-with-type.rdf"
              (build-ore agg-uri arch archived-files [{:attr "datacite.resourcetype" :value "The Type"}
                                                      {:attr "datacite.foo" :value "Bar"}]))
    (test-ore "ore-with-type.rdf"
              (build-ore agg-uri arch archived-files [{:attr "datacite.resourcetype" :value "The Type"}
                                                      {:attr "datacite.title" :value ""}]))))

(deftest test-ore-with-contributor
  (testing "ORE with contributor"
    (test-ore "ore-with-contributor.rdf"
              (build-ore agg-uri arch archived-files [{:attr "contributorName" :value "The Contributor"}]))
    (test-ore "ore-with-contributor.rdf"
              (build-ore agg-uri arch archived-files [{:attr "contributorName" :value "The Contributor"}
                                                      {:attr "datacite.foo" :value "Bar"}]))
    (test-ore "ore-with-contributor.rdf"
              (build-ore agg-uri arch archived-files [{:attr "contributorName" :value "The Contributor"}
                                                      {:attr "datacite.title" :value ""}]))))

(deftest test-ore-with-subject
  (testing "ORE with subject."
    (test-ore "ore-with-subject.rdf"
              (build-ore agg-uri arch archived-files [{:attr "Subject" :value "The Subject"}]))
    (test-ore "ore-with-subject.rdf"
              (build-ore agg-uri arch archived-files [{:attr "Subject" :value "The Subject"}
                                                      {:attr "datacite.foo" :value "Bar"}]))
    (test-ore "ore-with-subject.rdf"
              (build-ore agg-uri arch archived-files [{:attr "Subject" :value "The Subject"}
                                                      {:attr "datacite.title" :value ""}]))))

(deftest test-ore-with-rights
  (testing "ORE with rights."
    (test-ore "ore-with-rights.rdf"
              (build-ore agg-uri arch archived-files [{:attr "Rights" :value "The Rights"}]))
    (test-ore "ore-with-rights.rdf"
              (build-ore agg-uri arch archived-files [{:attr "Rights" :value "The Rights"}
                                                      {:attr "datacite.foo" :value "Bar"}]))
    (test-ore "ore-with-rights.rdf"
              (build-ore agg-uri arch archived-files [{:attr "Rights" :value "The Rights"}
                                                      {:attr "datacite.title" :value ""}]))))

(deftest test-ore-with-description
  (testing "ORE with description."
    (test-ore "ore-with-description.rdf"
              (build-ore agg-uri arch archived-files [{:attr "Description" :value "The Description"}]))
    (test-ore "ore-with-description.rdf"
              (build-ore agg-uri arch archived-files [{:attr "Description" :value "The Description"}
                                                      {:attr "datacite.foo" :value "Bar"}]))
    (test-ore "ore-with-description.rdf"
              (build-ore agg-uri arch archived-files [{:attr "Description" :value "The Description"}
                                                      {:attr "datacite.title" :value ""}]))))

(deftest test-ore-with-identifier
  (testing "ORE with identifier."
    (test-ore "ore-with-identifier.rdf"
              (build-ore agg-uri arch archived-files [{:attr "Identifier" :value "The Identifier"}]))
    (test-ore "ore-with-identifier.rdf"
              (build-ore agg-uri arch archived-files [{:attr "Identifier" :value "The Identifier"}
                                                      {:attr "datacite.foo" :value "Bar"}]))
    (test-ore "ore-with-identifier.rdf"
              (build-ore agg-uri arch archived-files [{:attr "Identifier" :value "The Identifier"}
                                                      {:attr "datacite.title" :value ""}]))))

(deftest test-ore-with-box
  (testing "ORE with box."
    (test-ore "ore-with-box.rdf"
              (build-ore agg-uri arch archived-files [{:attr "geoLocationBox" :value "The Box"}]))
    (test-ore "ore-with-box.rdf"
              (build-ore agg-uri arch archived-files [{:attr "geoLocationBox" :value "The Box"}
                                                      {:attr "datacite.foo" :value "Bar"}]))
    (test-ore "ore-with-box.rdf"
              (build-ore agg-uri arch archived-files [{:attr "geoLocationBox" :value "The Box"}
                                                      {:attr "datacite.title" :value ""}]))))

(deftest test-ore-with-location
  (testing "ORE with location."
    (test-ore "ore-with-location.rdf"
              (build-ore agg-uri arch archived-files [{:attr "geoLocationPlace" :value "The Place"}]))
    (test-ore "ore-with-location.rdf"
              (build-ore agg-uri arch archived-files [{:attr "geoLocationPlace" :value "The Place"}
                                                      {:attr "datacite.foo" :value "Bar"}]))
    (test-ore "ore-with-location.rdf"
              (build-ore agg-uri arch archived-files [{:attr "geoLocationPlace" :value "The Place"}
                                                      {:attr "datacite.title" :value ""}]))))

(deftest test-ore-with-point
  (testing "ORE with point."
    (test-ore "ore-with-point.rdf"
              (build-ore agg-uri arch archived-files [{:attr "geoLocationPoint" :value "The Point"}]))
    (test-ore "ore-with-point.rdf"
              (build-ore agg-uri arch archived-files [{:attr "geoLocationPoint" :value "The Point"}
                                                      {:attr "datacite.foo" :value "Bar"}]))
    (test-ore "ore-with-point.rdf"
              (build-ore agg-uri arch archived-files [{:attr "geoLocationPoint" :value "The Point"}
                                                      {:attr "datacite.title" :value ""}]))))


src/org/cyverse/oai_ore.clj
(ns org.cyverse.oai-ore
  (:use [clojure.data.xml :only [alias-uri element]])
  (:require [clojure.string :as string]))

(defprotocol RdfSerializable
  (to-rdf [_] "Serializes the object as RDF/XML. This function returns an instance of clojure.data.xml.Element."))

;; These are the namespaces that may be used in the the RDF serialization of the archive. The key is the abbreviation
;; as it should appear in the serialized XML. The value is the namespace URI.
(def ^:private namespaces
  {:cito    "http://purl.org/spar/cito/"
   :dc      "http://purl.org/dc/elements/1.1/"
   :dcterms "http://purl.org/dc/terms/"
   :foaf    "http://xmlns.com/foaf/0.1/"
   :ore     "http://www.openarchives.org/ore/terms/"
   :owl     "http://www.w3.org/2002/07/owl#"
   :rdf     "http://www.w3.org/1999/02/22-rdf-syntax-ns#"
   :rdfs    "http://www.w3.org/2000/01/rdf-schema#"})

;; Alias the namespaces so that they can easily be used when generating the XML.
(apply alias-uri (apply concat namespaces))

(defn- default-attribute-formatter
  "Returns a default attribute formatter for the given XML tag."
  [tag]
  (fn [content] (element tag {} content)))

(defn- default-dcterm-attribute-formatter
  "Returns a default formatter for dcterm attributes."
  [tag parse-type]
  (fn [content] (element tag {::rdf/parseType parse-type} content)))

;; A table that associates datacite terms with attribute names.
(def ^:private attribute-formatter-for
  {"datacite.title"        (default-attribute-formatter ::dc/title)
   "datacite.publisher"    (default-attribute-formatter ::dc/publisher)
   "datacite.creator"      (default-attribute-formatter ::dc/creator)
   "datacite.resourcetype" (default-attribute-formatter ::dc/type)
   "contributorName"       (default-attribute-formatter ::dc/contributor)
   "Subject"               (default-attribute-formatter ::dc/subject)
   "Rights"                (default-attribute-formatter ::dc/rights)
   "Description"           (default-attribute-formatter ::dc/description)
   "Identifier"            (default-attribute-formatter ::dcterms/identifier)
   "geoLocationBox"        (default-dcterm-attribute-formatter ::dcterms/Box "Literal")
   "geoLocationPlace"      (default-dcterm-attribute-formatter ::dcterms/Location "Literal")
   "geoLocationPoint"      (default-dcterm-attribute-formatter ::dcterms/Point "Literal")})

(defn- aggregates-element
  "Gnereates an RDF/XML element indicating that a file is contained within an aggregation."
  [file-uri]
  (when file-uri (element ::ore/aggregates {::rdf/resource file-uri})))

(defn- element-for
  "Generates an RDF/XML element for an AVU."
  [{:keys [attr value]}]
  (when-not (string/blank? value)
    (when-let [formatter (attribute-formatter-for attr)]
      (formatter (string/trim value)))))

(deftype Aggregation [uri file-uris avus meta-uri]
  RdfSerializable
  (to-rdf [_]
    (element ::rdf/Description {::rdf/about uri}
      (->> (concat [(element ::rdf/type {::rdf/resource "http://www.openarchives.org/ore/terms/Aggregation"})]
                   (mapv aggregates-element (concat [meta-uri] file-uris))
                   (map element-for avus))
           (remove nil?)
           doall))))

(deftype Archive [id uri aggregation-uri]
  RdfSerializable
  (to-rdf [_]
    (element ::rdf/Description {::rdf/about uri}
      [(element ::dcterms/identifier {} id)
       (element ::rdf/type {::rdf/resource "http://www.openarchives.org/ore/terms/ResourceMap"})
       (element ::ore/describes {::rdf/resource aggregation-uri})])))

(deftype MetadataFile [id meta-uri file-uris]
  RdfSerializable
  (to-rdf [_]
    (element ::rdf/Description {::rdf/about meta-uri}
      (concat [(element ::dcterms/identifier {} id)]
              (mapv (fn [{:keys [uri]}] (element ::cito/documents {::rdf/resource uri})) file-uris)))))

(deftype ArchivedFile [id file-uri meta-uri]
  RdfSerializable
  (to-rdf [_]
    (element ::rdf/Description {::rdf/about file-uri}
      (remove nil? [(element ::dcterms/identifier {} id)
                    (when meta-uri (element ::cito/isDocumentedBy {::rdf/resource meta-uri}))]))))

(deftype Ore [descriptions]
  RdfSerializable
  (to-rdf [_]
    (element ::rdf/RDF (into {} (map (fn [[k v]] [(keyword "xmlns" (name k)) v]) namespaces))
      (mapv to-rdf descriptions))))

(defn build-ore
  "Generates an ORE archive for a data set."
  [aggregation-uri archive archived-files & [avus metadata]]
  (Ore. (concat (remove nil? [(Aggregation. aggregation-uri (mapv :uri archived-files) avus (:uri metadata))
                              (Archive. (:id archive) (:uri archive) aggregation-uri)
                              (when metadata (MetadataFile. (:id metadata) (:uri metadata) archived-files))])
                (mapv (fn [{:keys [id uri]}] (ArchivedFile. id uri (:uri metadata))) archived-files))))

(def format-id "http://www.openarchives.org/ore/terms")
